pkg/acl/acl.go

```diff
@@ -0,0 +1,408 @@
+// {
+//   int (*acl_set_file)(const char *, acl_type_t, acl_t);
+//
+//   acl_set_file = f;
+//   return acl_set_file(path_p, type, acl);
+// }
+//
+// int
+// my_acl_free(void *f, acl_t acl)
+// {
+//   int (*acl_free)(acl_t);
+//
+//   acl_free = f;
+//   return acl_free(acl);
+// }
+//
+// int
+// my_acl_valid(void *f, acl_t acl)
+// {
+//   int (*acl_valid)(acl_t);
+//
+//   acl_valid = f;
+//   return acl_valid(acl);
+// }
+//
+// int
+// my_acl_create_entry(void *f, acl_t *acl_p, acl_entry_t *entry_p)
+// {
+//   int (*acl_create_entry)(acl_t *, acl_entry_t *);
+//
+//   acl_create_entry = f;
+//   return acl_create_entry(acl_p, entry_p);
+// }
+//
+// int
+// my_acl_set_tag_type(void *f, acl_entry_t entry_d, acl_tag_t tag_type)
+// {
+//   int (*acl_set_tag_type)(acl_entry_t, acl_tag_t);
+//
+//   acl_set_tag_type = f;
+//   return acl_set_tag_type(entry_d, tag_type);
+// }
+//
+// int
+// my_acl_get_permset(void *f, acl_entry_t entry_d, acl_permset_t *permset_p)
+// {
+//   int (*acl_get_permset)(acl_entry_t, acl_permset_t *);
+//
+//   acl_get_permset = f;
+//   return acl_get_permset(entry_d, permset_p);
+// }
+//
+// int
+// my_acl_add_perm(void *f, acl_permset_t permset_d, acl_perm_t perm)
+// {
+//   int (*acl_add_perm)(acl_permset_t, acl_perm_t);
+//
+//   acl_add_perm = f;
+//   return acl_add_perm(permset_d, perm);
+// }
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"unsafe"
+)
+
+const (
+	otherExec = 1 << iota
+	otherWrite
+	otherRead
+	groupExec
+	groupWrite
+	groupRead
+	userExec
+	userWrite
+	userRead
+)
+
+const (
+	TagUserObj  Tag = C.ACL_USER_OBJ
+	TagUser         = C.ACL_USER
+	TagGroupObj     = C.ACL_GROUP_OBJ
+	TagGroup        = C.ACL_GROUP
+	TagMask         = C.ACL_MASK
+	TagOther        = C.ACL_OTHER
+)
+
+const (
+	PermRead    Perm = C.ACL_READ
+	PermWrite        = C.ACL_WRITE
+	PermExecute      = C.ACL_EXECUTE
+)
+
+type ACL struct {
+	lib *libHandle
+	a   C.acl_t
+}
+
+// Entry is an entry in an ACL.
+type Entry struct {
+	a *ACL
+	e C.acl_entry_t
+}
+
+// Permset is a set of permissions.
+type Permset struct {
+	a *ACL
+	p C.acl_permset_t
+}
+
+// Perm represents a permission in the e_perm ACL field
+type Perm int
+
+// Tag represents an ACL e_tag entry
+type Tag int
+
+var ErrSoNotFound = errors.New("unable to open a handle to libacl")
+
+type libHandle struct {
+	handle  unsafe.Pointer
+	libname string
+}
+
+func getHandle() (*libHandle, error) {
+	for _, name := range []string{
+		"libacl.so.1",
+		"libacl.so",
+	} {
+		libname := C.CString(name)
+		defer C.free(unsafe.Pointer(libname))
+		handle := C.dlopen(libname, C.RTLD_LAZY)
+		if handle != nil {
+			h := &libHandle{
+				handle:  handle,
+				libname: name,
+			}
+			return h, nil
+		}
+	}
+	return nil, ErrSoNotFound
+}
+
+func getSymbolPointer(handle unsafe.Pointer, symbol string) (unsafe.Pointer, error) {
+	sym := C.CString(symbol)
+	defer C.free(unsafe.Pointer(sym))
+
+	p := C.dlsym(handle, sym)
```

Comment:
OK. I guess I'll wrap all the errors then.

BTW, we should also do this in [go-systemd](https://github.com/coreos/go-systemd/blob/2ed5b5012ccde5f057c197890a2c801295941149/util/util.go)